File: ledger-models-java/src/test/java/bench/LazyHydratePerfGuard.java
package bench;

import common.models.transaction.Transaction;
import common.util.LinkCache;
import fintekkers.models.portfolio.PortfolioProto;
import fintekkers.models.transaction.TransactionProto;
import fintekkers.models.util.LocalTimestamp.LocalTimestampProto;
import fintekkers.models.util.Uuid.UUIDProto;
import org.junit.jupiter.api.Test;
import protos.serializers.util.proto.ProtoSerializationUtil;

import java.time.ZonedDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Performance-regression guard for the Transaction wrapper's lazy-hydrate
 * read path (pre-warmed cache hit).
 *
 * <p>Baseline measured on a Mac Mini M-series: ~1.06 µs/op at N=10000 in
 * steady state (post-JIT). Default ceiling is the baseline + 15% headroom.
 *
 * <p>Override via {@code -DLAZY_HYDRATE_PERF_CEILING_US=<float>} on slower
 * CI hardware (or {@code LAZY_HYDRATE_PERF_CEILING_US} env var). Trade-off
 * documented in the perf bench commit: this is a wrapper-overhead guard,
 * not an end-to-end gRPC / DB latency guard.
 */
class LazyHydratePerfGuard {

    private static final int N = 10_000;
    private static final double DEFAULT_CEILING_US = 1.22; // 1.06 * 1.15

    @Test
    void per_op_stays_within_15pct_of_baseline_at_n_10000() {
        double ceilingUs = readCeiling();

        // Warmup: same loop, discard timing. Lets the JIT compile getTradeName /
        // ensureHydrated / LinkCache.get before we measure.
        runOnce();
        // Measure
        double perOpUs = runOnce();

        System.out.printf(
                "LazyHydratePerfGuard: N=%d  per_op=%.2f us  ceiling=%.2f us%n",
                N, perOpUs, ceilingUs);

        assertTrue(perOpUs <= ceilingUs,
                String.format(
                        "Transaction lazy-hydrate per-op (%.2f us) exceeded ceiling (%.2f us). "
                        + "Either a regression or noisy hardware; override via "
                        + "-DLAZY_HYDRATE_PERF_CEILING_US=<float> if running on slow CI.",
                        perOpUs, ceilingUs));
    }

    /** Build N link/resolved pairs, prime the cache, run the hot loop, return per-op µs. */
    private static double runOnce() {
        ZonedDateTime asOf = ZonedDateTime.now().withNano(0);
        LocalTimestampProto asOfProto = ProtoSerializationUtil.serializeTimestamp(asOf);

        List<UUID> txnUuids = new ArrayList<>(N);
        List<UUID> portUuids = new ArrayList<>(N);
        List<TransactionProto> linkProtos = new ArrayList<>(N);

        for (int i = 0; i < N; i++) {
            UUID txnUuid = UUID.randomUUID();
            UUID portUuid = UUID.randomUUID();
            txnUuids.add(txnUuid);
            portUuids.add(portUuid);

            UUIDProto txnUuidProto = ProtoSerializationUtil.serializeUUID(txnUuid);
            UUIDProto portUuidProto = ProtoSerializationUtil.serializeUUID(portUuid);

            PortfolioProto resolvedPort = PortfolioProto.newBuilder()
                    .setUuid(portUuidProto).setAsOf(asOfProto).setIsLink(false)
                    .setPortfolioName("P-" + portUuid.toString().substring(0, 8)).build();
            TransactionProto resolved = TransactionProto.newBuilder()
                    .setUuid(txnUuidProto).setAsOf(asOfProto).setIsLink(false)
                    .setTradeName("T-" + txnUuid.toString().substring(0, 8))
                    .setPortfolio(resolvedPort).build();
            LinkCache.TRANSACTION.put(txnUuid, resolved, asOf);
            LinkCache.PORTFOLIO.put(portUuid, resolvedPort, asOf);

            linkProtos.add(TransactionProto.newBuilder()
                    .setUuid(txnUuidProto).setAsOf(asOfProto).setIsLink(true).build());
        }

        System.gc();
        long t0 = System.nanoTime();
        int sink = 0;
        for (TransactionProto link : linkProtos) {
            Transaction txn = new Transaction(link);
            String name = txn.getTradeName();
            if (name != null && !name.isEmpty()) sink++;
        }
        long t1 = System.nanoTime();
        // Sink the read so the JIT can't elide the work.
        if (sink < 0) throw new AssertionError();

        for (UUID u : txnUuids) LinkCache.TRANSACTION.evict(u);
        for (UUID u : portUuids) LinkCache.PORTFOLIO.evict(u);

        return ((double) (t1 - t0) / N) / 1_000.0;
    }

    private static double readCeiling() {
        String fromProp = System.getProperty("LAZY_HYDRATE_PERF_CEILING_US");
        String fromEnv = System.getenv("LAZY_HYDRATE_PERF_CEILING_US");
        String raw = fromProp != null ? fromProp : fromEnv;
        if (raw == null) return DEFAULT_CEILING_US;
        try {
            return Double.parseDouble(raw);
        } catch (NumberFormatException e) {
            return DEFAULT_CEILING_US;
        }
    }
}

File: ledger-models-rust/tests/lazy_hydrate_perf_guard.rs
//! Performance-regression guard for TransactionWrapper lazy-hydrate
//! (pre-warmed cache hit). Mirrors LazyHydratePerfGuard.java.
//!
//! Baseline (release, Mac Mini M-series): ~0.82 µs/op at N=10000 in steady
//! state. Default ceiling = baseline + 15% headroom = 0.94 µs/op.
//!
//! Debug builds skip the strict assertion (Rust debug is unoptimized and
//! 10–50× slower than release). Run with:
//!
//!   cargo test --release --test lazy_hydrate_perf_guard -- --nocapture
//!
//! Override via env var LAZY_HYDRATE_PERF_CEILING_US for slow CI hardware.

use ledger_models::fintekkers::models::portfolio::PortfolioProto;
use ledger_models::fintekkers::models::transaction::TransactionProto;
use ledger_models::fintekkers::models::util::{LocalTimestampProto, UuidProto};
use ledger_models::fintekkers::wrappers::models::transaction::TransactionWrapper;
use ledger_models::fintekkers::wrappers::util::link_cache;
use prost_types::Timestamp;
use std::time::Instant;
use uuid::Uuid;

const N: usize = 10_000;
const DEFAULT_CEILING_US: f64 = 0.94; // 0.82 * 1.15

fn read_ceiling_us() -> f64 {
    match std::env::var("LAZY_HYDRATE_PERF_CEILING_US")
        .ok()
        .and_then(|s| s.parse::<f64>().ok())
    {
        Some(v) => v,
        None => DEFAULT_CEILING_US,
    }
}

fn make_as_of() -> LocalTimestampProto {
    LocalTimestampProto {
        timestamp: Some(Timestamp {
            seconds: 1_700_000_000,
            nanos: 0,
        }),
        time_zone: "UTC".to_string(),
    }
}

fn run_once() -> f64 {
    let as_of = make_as_of();

    let mut txn_uuids: Vec<Uuid> = Vec::with_capacity(N);
    let mut port_uuids: Vec<Uuid> = Vec::with_capacity(N);
    let mut links: Vec<TransactionProto> = Vec::with_capacity(N);

    for _ in 0..N {
        let txn_uuid = Uuid::new_v4();
        let port_uuid = Uuid::new_v4();
        txn_uuids.push(txn_uuid);
        port_uuids.push(port_uuid);

        let resolved_port = PortfolioProto {
            uuid: Some(UuidProto {
                raw_uuid: port_uuid.as_bytes().to_vec(),
            }),
            as_of: Some(as_of.clone()),
            is_link: false,
            portfolio_name: format!("P-{}", &port_uuid.to_string()[..8]),
            ..Default::default()
        };
        let resolved = TransactionProto {
            uuid: Some(UuidProto {
                raw_uuid: txn_uuid.as_bytes().to_vec(),
            }),
            as_of: Some(as_of.clone()),
            is_link: false,
            trade_name: format!("T-{}", &txn_uuid.to_string()[..8]),
            portfolio: Some(resolved_port.clone()),
            ..Default::default()
        };
        link_cache::transaction().put(txn_uuid, resolved, Some(as_of.clone()));
        link_cache::portfolio().put(port_uuid, resolved_port, Some(as_of.clone()));

        links.push(TransactionProto {
            uuid: Some(UuidProto {
                raw_uuid: txn_uuid.as_bytes().to_vec(),
            }),
            as_of: Some(as_of.clone()),
            is_link: true,
            ..Default::default()
        });
    }

    let t0 = Instant::now();
    let mut sink = 0usize;
    for link in &links {
        let w = TransactionWrapper::new(link.clone());
        if !w.trade_name().is_empty() {
            sink += 1;
        }
    }
    let elapsed = t0.elapsed();
    assert_eq!(sink, N);

    for u in &txn_uuids {
        link_cache::transaction().evict(*u);
    }
    for u in &port_uuids {
        link_cache::portfolio().evict(*u);
    }

    elapsed.as_nanos() as f64 / N as f64 / 1_000.0
}

#[test]
fn per_op_stays_within_15pct_of_baseline_at_n_10000() {
    let ceiling_us = read_ceiling_us();

    // Warmup
    run_once();
    let per_op_us = run_once();

    let mode = if cfg!(debug_assertions) { "debug" } else { "release" };
    println!(
        "LazyHydratePerfGuard (rust/{}): N={}  per_op={:.2} us  ceiling={:.2} us",
        mode, N, per_op_us, ceiling_us
    );

    if cfg!(debug_assertions) {
        // Debug builds are unoptimized — log only. Use --release for the
        // real assertion.
        eprintln!(
            "(debug build — strict perf check skipped; re-run with --release)"
        );
        return;
    }

    assert!(
        per_op_us <= ceiling_us,
        "Transaction lazy-hydrate per-op ({:.2} us) exceeded ceiling ({:.2} us). \
         Either a regression or noisy hardware; override via \
         LAZY_HYDRATE_PERF_CEILING_US env var if running on slow CI.",
        per_op_us,
        ceiling_us
    );
}
